internal/vulnstore/postgres/gc.go

```diff
@@ -0,0 +1,265 @@
+package postgres
+
+import (
+	"context"
+	"fmt"
+	"runtime"
+
+	"github.com/google/uuid"
+	"github.com/jackc/pgtype"
+	"github.com/jackc/pgx/v4"
+	"github.com/jackc/pgx/v4/pgxpool"
+	"golang.org/x/sync/semaphore"
+)
+
+// GC is split into two phases, first it will identify any update operations
+// which are older then the provided keep value and delete these.
+//
+// Next it will perform chunked deletions of any vulns from the vuln table
+// which are not longer referenced by update operations.
+//
+// The GC is throttled to not overload the database with cascade deletes.
+// If a full GC is required run this method until the returned int64 value
+// is 0.
+func (s *Store) GC(ctx context.Context, keep int) (int64, error) {
+	const (
+		// GCThrottle sets a limit for the number of deleted update operations
+		// (and subsequent cascade deletes in the uo_vuln table) that can occur in a GC run.
+		GCThrottle = 50
+	)
+
+	// obtain update operations which need deletin'
+	ops, totalOps, err := eligibleUpdateOpts(ctx, s.pool, keep)
+	if err != nil {
+		return 0, err
+	}
+
+	// delete em', but not too many...
+	if totalOps >= GCThrottle {
+		ops = ops[:GCThrottle]
+	}
+
+	deletedOps, err := s.DeleteUpdateOperations(ctx, ops...)
+	if err != nil {
+		return totalOps - deletedOps, err
+	}
+
+	// get all updaters we know about.
+	updaters, err := distinctUpdaters(ctx, s.pool)
+
+	// issue concurrent chunk'd deletion for known updaters
+	// limit concurrency by available CPUs.
+	sem := semaphore.NewWeighted(int64(runtime.NumCPU()))
+
+	errC := make(chan error, len(updaters))
+
+	for _, updater := range updaters {
+		err = sem.Acquire(ctx, 1)
+		if err != nil {
+			break
+		}
+		go func(u string) {
+			defer sem.Release(1)
+			err := chunkedCleanup(ctx, s.pool, u)
+			if err != nil {
+				errC <- err
+			}
+		}(updater)
+	}
+
+	// unconditionally wait for all in-flight go routines to return.
+	// the use of context.Background and lack of error checking is intentional.
+	// all in-flight go routines are guarantee to release their sems.
+	sem.Acquire(context.Background(), int64(runtime.NumCPU()))
+
+	close(errC)
+	for e := range errC {
+		if e != nil {
+			return totalOps - deletedOps, fmt.Errorf("received error issuing deletes: %v", err)
```

Comment:
This only ever returns the first error in the channel, which isn't _necessarily_ the first error encountered. I think it'd be worth the cycles to build a comprehensive error message in this path.

Reply:
Do we care about "first" error, since they can all run in parallel? I'm not necessary seeing why we care about "first" error - order isn't really of concern AFAICT.

Reply:
Right, but as it stands, this just returns one of the errors, not all of them. The first one to occur is probably going to have the most informative error, but this just grabs a random one.

Reply:
Alright, probably easiest to round up all the errors into a channel then, and build an error string like we do with updaters. 

Dont want to add a sync.Once here if i can avoid it.